File: built-in plugins dsp/mb-restore-declip.cpp
/**
 * MB De-Clip
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Reconstruct clipped audio peaks for damaged recordings
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_DECLIP_H
#define MB_RESTORE_DECLIP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreDeclip : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-declip";
    static constexpr const char* PLUGIN_NAME    = "MB De-Clip";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -1f;  // range [-6, 0]
    float quality = 2f;  // range [0, 3]
    float makeup = 0f;  // range [-12, 12]
    };

    MbRestoreDeclip() = default;
    ~MbRestoreDeclip() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -6f, 0f);
        params.quality = std::clamp(params.quality, 0f, 3f);
        params.makeup = std::clamp(params.makeup, -12f, 12f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB De-Clip
        return input;
    }
};

#endif // MB_RESTORE_DECLIP_H

File: built-in plugins dsp/mb-restore-declip.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbRestoreDeclipPlugin: PluginDefinition = {
  id: "mb-restore-declip",
  slug: "mb-restore-declip",
  name: "MB De-Clip",
  category: "effect",
  type: "distortion" as any,
  version: "1.0.0",
  description: "Reconstruct clipped audio peaks for damaged recordings",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "threshold",
      name: "Threshold",
      type: "float",
      defaultValue: -1,
      minValue: -6,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "quality",
      name: "Quality",
      type: "float",
      defaultValue: 2,
      minValue: 0,
      maxValue: 3,
      automatable: false,
    },
    {
      id: "makeup",
      name: "Makeup Gain",
      type: "float",
      defaultValue: 0,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
  ],
  defaultPreset: { threshold: -1, quality: 2, makeup: 0 },
};

export default MbRestoreDeclipPlugin;
